pvlib/modelchain.py

```diff
@@ -931,23 +996,125 @@ def prepare_inputs(self, weather, times=None):
+            speed of 0 m/s are assumed.
+
+        Notes
+        -----
+        Assigns attributes: ``weather``, ``total_irrad``, ``solar_position``,
+        ``airmass``, ``aoi``, ``temperature``.
+
+        See also
+        --------
+        pvlib.modelchain.ModelChain.prepare_inputs
+        """
+
+        self._assign_weather(data)
+
+        self._verify_df(data, req=['poa_global', 'poa_direct', 'poa_diffuse'])
+        self._assign_total_irrad(data)
+
+        self._prep_inputs_solar_pos()
+        self._prep_inputs_airmass()
+
+        if isinstance(self.system, SingleAxisTracker):
+            self._prep_inputs_tracking()
+        else:
+            self._prep_inputs_fixed()
+
+        return self
+
+    def prepare_temperature(self, data=None):
+        """
+        Sets cell_temperature using inputs in data and the specified
+        temperature model.
+
+        If 'data' contains 'cell_temperature', these values are assigned to
+        attribute ``cell_temperature``. If 'data' contains 'module_temperature`
+        and `temperature_model' is 'sapm', cell temperature is calculated using
+        :py:func:`pvlib.temperature.sapm_celL_from_module`. Otherwise, cell
+        temperature is calculated by 'temperature_model'.
+
+        Parameters
+        ----------
+        data : DataFrame, default None
+            May contain columns ``'cell_temperature'`` or
+            ``'module_temperaure'``.
+
+        Returns
+        -------
+        self
+
+        Assigns attribute ``cell_temperature``.
+
+        """
+        try:
+            self.cell_temperature = data['cell_temperature']
+            if self.temperature_model is not None:
+                warnings.warn('using ''cell_temperature'' provided in input',
+                              ' rather than specified temperature model: {0}'
+                              .format(self.temperature_model.__name__))
+            return self
+        except:
+            pass
+
+        # cell_temperature is not in input. Calculate cell_temperature using
+        # a temperature_model.
+        # If module_temperature is in input data we can use the SAPM cell
+        # temperature model.
+        if ('module_temperature' in data) and \
+            (self.temperature_model.__name__ == 'sapm'):
+            # use SAPM cell temperature model only
+            self.cell_temperature = pvlib.temperature.sapm_cell_from_module(
+                module_temperature=data['module_temperature'],
+                poa_global=self.total_irrad['poa_global'],
+                deltaT=self.temperature_model_parameters['deltaT'])
+            return self
+
+        # Calculate cell temperature from weather data. Cell temperature models
+        # expect total_irrad['poa_global']. If not found, substitute
+        # self.effective_irradiance.
+        if 'poa_global' not in self.total_irrad:
+            try:
+                self.total_irrad['poa_global'] = self.effective_irradiance
```

Comment:
The problem is that the ModelChain cell temperature methods reference `ModelChain.total_irrad['poa_global']` as the irradiance input to the cell temperature function.  I could change `ModelChain.prepare_temperature` to expect that `ModelChain.total_irrad['poa_global']` has already been assigned, but, this means that `ModelChain.run_from_effective_irradiance` requires data (poa_global) that is upstream of `effective_irradiance` and which the user may not have (poa_global).

I would prefer not to assign `effective_irradiance` to `total_irrad['poa_global']` because that's not what 'poa_global' means. Maybe have the cell temperature methods fall back to `ModelChain.effective_irradiance` if 'poa_global' is not found?

Reply:
> Maybe have the cell temperature methods fall back to ModelChain.effective_irradiance if 'poa_global' is not found?

Good idea. Would that fallback need to emit a warning? I'd lean towards no but I don't have a strong opinion about it.

Reply:
Were it my analysis I would appreciate the warning that pvlib swapped the irradiance inputs for me.

Let's take this up in a follow-on PR since it will rework the cell temperature methods, rather than the `run_model` methods.